src/smartstream/src/lib.rs

```diff
@@ -0,0 +1,78 @@
+use std::fmt;
+pub use fluvio_dataplane_protocol as dataplane;
+use dataplane::record::DefaultRecord;
+
+#[cfg(feature = "derive")]
+pub use fluvio_smartstream_derive::smartstream;
+
+pub mod memory {
+    /// Allocate memory into the module's linear memory
+    /// and return the offset to the start of the block.
+    #[no_mangle]
+    pub fn alloc(len: usize) -> *mut u8 {
+        // create a new mutable buffer with capacity `len`
+        let mut buf = Vec::with_capacity(len);
+        // take a mutable pointer to the buffer
+        let ptr = buf.as_mut_ptr();
+        // take ownership of the memory block and
+        // ensure the its destructor is not
+        // called when the object goes out of scope
+        // at the end of the function
+        std::mem::forget(buf);
+        // return the pointer so the runtime
+        // can write data at this offset
+        ptr
+    }
+
+    #[no_mangle]
+    #[allow(clippy::missing_safety_doc)]
+    pub unsafe fn dealloc(ptr: *mut u8, size: usize) {
+        let data = Vec::from_raw_parts(ptr, size, size);
+
+        std::mem::drop(data);
+    }
+}
+
+/// A trait for any type that may construct itself from a DefaultRecord.
+pub trait FromRecord: Sized {
+    type Error: std::error::Error;
+    fn from_record(record: &DefaultRecord) -> Result<Self, Self::Error>;
+}
+
+pub trait IntoRecord {
+    fn into_record(self) -> (Option<Vec<u8>>, Vec<u8>);
+}
+
+#[derive(Debug)]
+pub enum VoidError {}
+impl fmt::Display for VoidError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        writeln!(f, "This error is impossible")
+    }
+}
+impl std::error::Error for VoidError {}
+
+#[derive(Debug)]
+pub struct SimpleRecord {
```

Comment:
Why creating another abstraction instead of using DefaultRecord?

Reply:
DefaultRecord is just Record<DefaultAsyncBuffer>, and the Record type also contains data like the preamble, which contains fields that seem to be internal to the protocol and should probably not be edited by users. So SimpleRecord represents all the data from DefaultRecord that we actually want users to be able to interact with

Reply:
My previous feedback was that we can certainly hide preamble field which should issue.  We should not create unnecessary abstraction when we don't know what this will be in final form.  

Also, some of the preamble will be needed for time based  filtering.